From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.15 october2005, patch 02-40

  Description: qeth: tx_bytes and rx_bytes counter aren't set properly
  Symptom:     TX and RX byte counter are wrong when comparing them with
               VM counter.
  Problem:     qeth driver has also count the qdio header size for each
               skb.
  Solution:    don't add the qdio header size to TX and RX counter. Just count
               the payload without any hardware header size.
  Problem-ID:  
  Archs:       s390-31, s390-64

Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.5/drivers/s390/net/qeth_main.c	2 Feb 2006 18:03:13 -0000	1.252.2.3
+++ linux-2.5/drivers/s390/net/qeth_main.c	19 Mar 2006 19:42:36 -0000	1.252.2.4
@@ -4422,6 +4422,7 @@
 	int elements_needed = 0;
 	enum qeth_large_send_types large_send = QETH_LARGE_SEND_NO;
 	struct qeth_eddp_context *ctx = NULL;
+	int tx_bytes = skb->len;
 	int rc;
 
 	QETH_DBF_TEXT(trace, 6, "sendpkt");
@@ -4502,7 +4503,7 @@
 					      elements_needed, ctx);
 	if (!rc){
 		card->stats.tx_packets++;
-		card->stats.tx_bytes += skb->len;
+		card->stats.tx_bytes += tx_bytes;
 #ifdef CONFIG_QETH_PERF_STATS
 		if (skb_shinfo(skb)->tso_size &&
 		   !(large_send == QETH_LARGE_SEND_NO)) {

